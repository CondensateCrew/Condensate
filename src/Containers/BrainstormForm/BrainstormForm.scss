@import '../../css/variables.scss';
@import '../../css/index.scss';

.brainstorm-form {
  padding: $big-block-padding-space;
  background: $primary-gradient;
  border-radius: $big-blocks-border-radius;
  box-shadow: $external-shadow;
  overflow: scroll;
  h2 {
    font-size: $medium-text;
  }
}

.brainstorm-div {
  text-align: left;
  margin: $large-space 0;
  label {
    font-size: $small-text;
    font-weight: 500;
    text-transform: uppercase;
    margin: 0;
    display: -webkit-flex;
    display: -ms-flex;
    display: flex;
    align-items: center;
    justify-content: space-between;
    position: relative;
    z-index: 20;
    img {
      cursor: pointer;
      width: $small-text;
      margin-right: $small-space;
      &:hover {
        transform: scale(1.1);
      }
      &:active {
        transform: scale(0.9);
      }
    }
  }
  textarea {
    @extend %regular-input;
    border-radius: $one-line-border-radius;
    resize: none;
    padding: $small-space;
    height: 6rem;
  }
}

.brainstorm-form-menu {
  @extend %regular-input;
  display: flex;
  justify-content: space-between;
  align-items: center;
  border-radius: $one-line-border-radius;
  width: 100%;
  padding: $one-line-padding-space;
}

.cancel-btn {
  @extend %regular-btn;
  margin: 0;
  background: $error-color;
  font-weight: bold;
  &:hover {
    background: $darken-error-color;
  }
}

.clean-btn {
  display: -webkit-flex;
  display: -ms-flex;
  display: flex;
  align-items: center;
  justify-content: center;
  margin: 0;
  padding: $extra-small-space;
  border-radius: $round-border-radius;
  @extend %regular-btn;
  cursor: pointer;
  img {
    width: $small-text;
  }
}

.start-btn {
  margin: 0;
  @extend %regular-btn;
}
