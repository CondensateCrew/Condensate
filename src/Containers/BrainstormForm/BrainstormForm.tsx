import React, { useState, useEffect } from 'react';
import { useDispatch } from 'react-redux';
import './BrainstormForm.scss';
import broom from 'assets/broom.svg';
import down from 'assets/down.svg';
import IdeaQuestion from 'Components/IdeaQuestion/IdeaQuestion';
import CategoriesField from '../CategoriesField/CategoriesField';
import ActionsField from '../ActionsField/ActionsField';
import { IBrainstormForm, Category } from 'interfaces';
import { addCurrentBrainstorm } from 'redux/actions';
import { useHistory } from 'react-router-dom';

interface Props {
  brainstormFormState: boolean,
  cancel: (brainstormFormState:boolean) => void;
}
const BrainstormForm:React.FC<Props> = ({brainstormFormState, cancel }) => {
  const [ formState, setFormState ] = useState<IBrainstormForm>({
    question: '',
    categories: [],
    action: 'create an app',
    reset: false
  });
  const [ disabled, setIsDisabled ] = useState<boolean>(true);
  const [ error, setError ] = useState<string>('');
  const [ summary, setSummary ] = useState<string>('');
  const [ isClickedCategory, setIsClickedCategory ] = useState<boolean>(false);
  const dispatch = useDispatch();
  let history = useHistory();

  const resetForm = ():void => {
      setFormState({
        question: '',
        categories: [],
        action: 'create an app',
        reset: true
      })
      cancel(!brainstormFormState)
  }

  const validateFields = (state: IBrainstormForm) => {
    if (state.question.length > 0 && state.categories.length > 0) {
      return true
    }
      return false
  }

  const writeError = ():void => {
    if (formState.question === '') return setError('Please add a question');
    if (formState.categories === []) return setError('Please select a category');
  }

  const writeSummary = ():void => {
    let categoriesSum = formState.categories.map((ctg:Category) => {
      if (ctg.name === 'Environment') {
        return 'the environment'
      } else {
        return ctg.name.toLowerCase();
      }
    });

    setSummary(`I want to ${formState.action.toLowerCase()} about ${categoriesSum.join(', ')} to answer the question: ${formState.question}`)
  }

  const handleSubmit = ():void => {
    let currentBS = {
      id: 4,
      question: formState.question,
      idea: '',
      action: formState.action,
      isGenius: false,
      categories: formState.categories
    }
    validateFields(formState) ? dispatch(addCurrentBrainstorm(currentBS))
    : writeError();

    history.push('/instructions/first-rd');
  }

  const toggleCategory = (): void => setIsClickedCategory(!isClickedCategory);

  useEffect(() => {
      if (validateFields(formState)) {
        setIsDisabled(false)
        writeSummary();
      } else {
        setIsDisabled(true)
      }
  }, [formState])

  return (
    <form className='brainstorm-form'>
      {error !== '' && <p className="error-notification">{error}</p>}
      <h2 className='brainstorm-form-h2'>CREATE NEW BRAINSTORM</h2>
      <div className='brainstorm-div'>
        <label className='brainstorm-label'>IDEA QUESTION</label>
        <IdeaQuestion formState={formState} setQuestion={setFormState} />
      </div>
      <div className='brainstorm-div'>
        <label className='brainstorm-label'>CATEGORIES
          <img src={down} alt="down" onClick={toggleCategory}/>
        </label>
        {isClickedCategory && <CategoriesField formState={formState} setCategory={setFormState} />}
      </div>
      <div className='brainstorm-div'>
        <label className='brainstorm-label'>DESIRED ACTION</label>
        <ActionsField formState={formState} setAction={setFormState}/>
      </div>
      <div className='brainstorm-div'>
        <label className='brainstorm-label'>SUMMARY</label>
        <textarea readOnly className='brainstorm-summary-textarea' value={summary} placeholder='Typed summary here...'></textarea>
      </div>
      <div className='brainstorm-form-menu'>
        <button type='button' className='cancel-btn' onClick={resetForm}>cancel</button>
        <img className='broom' alt='broom-icon' src={broom}/>
        <button type='button' onClick={handleSubmit} disabled={disabled} className='start-btn'>start</button>
      </div>
    </form>
  )
}

export default BrainstormForm;
