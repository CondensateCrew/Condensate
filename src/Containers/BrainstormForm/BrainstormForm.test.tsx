/**
 * @jest-environment jsdom
 */

import React from 'react';
import { mount } from 'enzyme';
import BrainstormForm from './BrainstormForm';
import { MemoryRouter } from 'react-router-dom';

const mockDispatch = jest.fn();
jest.mock('react-redux', () => ({
  useDispatch: () => mockDispatch,
  useSelector: () => [{id: 1, action: 'Create an App'}]
}));

describe('BrainstormForm', () => {
  let wrapper: any;

  const mockBrainstormFormState = true;
  const mockCancel = jest.fn();

  beforeEach(() => {
    jest.mock('react-router-dom', () => ({
      useHistory: () => ({
        push: jest.fn(),
      }),
    }));

    wrapper = mount(<MemoryRouter
      initialEntries={[ { pathname: '/', key: 'testKey' } ]}><BrainstormForm
      brainstormFormState={mockBrainstormFormState}
      cancel={mockCancel}/></MemoryRouter>)
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should call cancel when the button is clicked', () => {
    wrapper.find('.cancel-btn').simulate('click');
    expect(mockCancel).toHaveBeenCalled();
  });

  it('should reset state when resetForm is called', () => {
    const mockEvent: Object = {target: {value: 'Who ate my cheeseburger?'}}
    wrapper.find('input').simulate('change', mockEvent);
    expect(wrapper).toMatchSnapshot();

    wrapper.find('button').at(1).simulate('click');
    expect(wrapper).toMatchSnapshot();
  });
});
