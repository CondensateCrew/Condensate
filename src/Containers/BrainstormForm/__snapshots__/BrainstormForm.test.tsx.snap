// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BrainstormForm should match the snapshot 1`] = `
<MemoryRouter
  initialEntries={
    Array [
      Object {
        "key": "testKey",
        "pathname": "/",
      },
    ]
  }
>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "canGo": [Function],
        "createHref": [Function],
        "entries": Array [
          Object {
            "hash": "",
            "key": "testKey",
            "pathname": "/",
            "search": "",
          },
        ],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "index": 0,
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "key": "testKey",
          "pathname": "/",
          "search": "",
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <BrainstormForm
      brainstormFormState={true}
      cancel={[MockFunction]}
    >
      <form
        className="brainstorm-form"
      >
        <h2
          className="brainstorm-form-h2"
        >
          CREATE NEW BRAINSTORM
        </h2>
        <div
          className="brainstorm-div"
        >
          <label
            className="brainstorm-label"
          >
            IDEA QUESTION
          </label>
          <IdeaQuestion
            formState={
              Object {
                "action": Object {
                  "action": "Create an App",
                  "id": 1,
                },
                "categories": Array [],
                "question": "",
                "reset": false,
              }
            }
            setQuestion={[Function]}
          >
            <input
              className="idea-question-input"
              onChange={[Function]}
              placeholder="Enter your idea here..."
              type="text"
              value=""
            />
          </IdeaQuestion>
        </div>
        <div
          className="brainstorm-div"
        >
          <label
            className="brainstorm-label"
          >
            CATEGORIES
            <img
              alt="down"
              onClick={[Function]}
              src="down.svg"
            />
          </label>
        </div>
        <div
          className="brainstorm-div"
        >
          <label
            className="brainstorm-label"
          >
            DESIRED ACTION
          </label>
          <ActionsField
            formState={
              Object {
                "action": Object {
                  "action": "Create an App",
                  "id": 1,
                },
                "categories": Array [],
                "question": "",
                "reset": false,
              }
            }
            setAction={[Function]}
          >
            <div
              className="action-toggle"
            >
              <header
                className=""
                onClick={[Function]}
              >
                <p
                  className="chosen"
                >
                  Create an app
                </p>
                <img
                  alt="down"
                  src="down.svg"
                />
              </header>
            </div>
          </ActionsField>
        </div>
        <div
          className="brainstorm-div"
        >
          <label
            className="brainstorm-label"
          >
            SUMMARY
          </label>
          <textarea
            className="brainstorm-summary-textarea"
            placeholder="Typed summary here..."
            readOnly={true}
            value=""
          />
        </div>
        <div
          className="brainstorm-form-menu"
        >
          <button
            className="cancel-btn"
            onClick={[Function]}
            type="button"
          >
            cancel
          </button>
          <button
            className="clean-btn"
            onClick={[Function]}
            type="button"
          >
            <img
              alt="broom-icon"
              className="broom"
              src="broom.svg"
            />
          </button>
          <button
            className="start-btn"
            disabled={true}
            onClick={[Function]}
            type="button"
          >
            start
          </button>
        </div>
      </form>
    </BrainstormForm>
  </Router>
</MemoryRouter>
`;

exports[`BrainstormForm should reset state when resetForm is called 1`] = `
<MemoryRouter
  initialEntries={
    Array [
      Object {
        "key": "testKey",
        "pathname": "/",
      },
    ]
  }
>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "canGo": [Function],
        "createHref": [Function],
        "entries": Array [
          Object {
            "hash": "",
            "key": "testKey",
            "pathname": "/",
            "search": "",
          },
        ],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "index": 0,
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "key": "testKey",
          "pathname": "/",
          "search": "",
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <BrainstormForm
      brainstormFormState={true}
      cancel={[MockFunction]}
    >
      <form
        className="brainstorm-form"
      >
        <h2
          className="brainstorm-form-h2"
        >
          CREATE NEW BRAINSTORM
        </h2>
        <div
          className="brainstorm-div"
        >
          <label
            className="brainstorm-label"
          >
            IDEA QUESTION
          </label>
          <IdeaQuestion
            formState={
              Object {
                "action": Object {
                  "action": "Create an App",
                  "id": 1,
                },
                "categories": Array [],
                "question": "Who ate my cheeseburger?",
                "reset": false,
              }
            }
            setQuestion={[Function]}
          >
            <input
              className="idea-question-input"
              onChange={[Function]}
              placeholder="Enter your idea here..."
              type="text"
              value="Who ate my cheeseburger?"
            />
          </IdeaQuestion>
        </div>
        <div
          className="brainstorm-div"
        >
          <label
            className="brainstorm-label"
          >
            CATEGORIES
            <img
              alt="down"
              onClick={[Function]}
              src="down.svg"
            />
          </label>
        </div>
        <div
          className="brainstorm-div"
        >
          <label
            className="brainstorm-label"
          >
            DESIRED ACTION
          </label>
          <ActionsField
            formState={
              Object {
                "action": Object {
                  "action": "Create an App",
                  "id": 1,
                },
                "categories": Array [],
                "question": "Who ate my cheeseburger?",
                "reset": false,
              }
            }
            setAction={[Function]}
          >
            <div
              className="action-toggle"
            >
              <header
                className=""
                onClick={[Function]}
              >
                <p
                  className="chosen"
                >
                  Create an app
                </p>
                <img
                  alt="down"
                  src="down.svg"
                />
              </header>
            </div>
          </ActionsField>
        </div>
        <div
          className="brainstorm-div"
        >
          <label
            className="brainstorm-label"
          >
            SUMMARY
          </label>
          <textarea
            className="brainstorm-summary-textarea"
            placeholder="Typed summary here..."
            readOnly={true}
            value=""
          />
        </div>
        <div
          className="brainstorm-form-menu"
        >
          <button
            className="cancel-btn"
            onClick={[Function]}
            type="button"
          >
            cancel
          </button>
          <button
            className="clean-btn"
            onClick={[Function]}
            type="button"
          >
            <img
              alt="broom-icon"
              className="broom"
              src="broom.svg"
            />
          </button>
          <button
            className="start-btn"
            disabled={true}
            onClick={[Function]}
            type="button"
          >
            start
          </button>
        </div>
      </form>
    </BrainstormForm>
  </Router>
</MemoryRouter>
`;

exports[`BrainstormForm should reset state when resetForm is called 2`] = `
<MemoryRouter
  initialEntries={
    Array [
      Object {
        "key": "testKey",
        "pathname": "/",
      },
    ]
  }
>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "canGo": [Function],
        "createHref": [Function],
        "entries": Array [
          Object {
            "hash": "",
            "key": "testKey",
            "pathname": "/",
            "search": "",
          },
        ],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "index": 0,
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "key": "testKey",
          "pathname": "/",
          "search": "",
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <BrainstormForm
      brainstormFormState={true}
      cancel={[MockFunction]}
    >
      <form
        className="brainstorm-form"
      >
        <h2
          className="brainstorm-form-h2"
        >
          CREATE NEW BRAINSTORM
        </h2>
        <div
          className="brainstorm-div"
        >
          <label
            className="brainstorm-label"
          >
            IDEA QUESTION
          </label>
          <IdeaQuestion
            formState={
              Object {
                "action": Object {
                  "action": "Create an App",
                  "id": 1,
                },
                "categories": Array [],
                "question": "",
                "reset": false,
              }
            }
            setQuestion={[Function]}
          >
            <input
              className="idea-question-input"
              onChange={[Function]}
              placeholder="Enter your idea here..."
              type="text"
              value=""
            />
          </IdeaQuestion>
        </div>
        <div
          className="brainstorm-div"
        >
          <label
            className="brainstorm-label"
          >
            CATEGORIES
            <img
              alt="down"
              onClick={[Function]}
              src="down.svg"
            />
          </label>
        </div>
        <div
          className="brainstorm-div"
        >
          <label
            className="brainstorm-label"
          >
            DESIRED ACTION
          </label>
          <ActionsField
            formState={
              Object {
                "action": Object {
                  "action": "Create an App",
                  "id": 1,
                },
                "categories": Array [],
                "question": "",
                "reset": false,
              }
            }
            setAction={[Function]}
          >
            <div
              className="action-toggle"
            >
              <header
                className=""
                onClick={[Function]}
              >
                <p
                  className="chosen"
                >
                  Create an app
                </p>
                <img
                  alt="down"
                  src="down.svg"
                />
              </header>
            </div>
          </ActionsField>
        </div>
        <div
          className="brainstorm-div"
        >
          <label
            className="brainstorm-label"
          >
            SUMMARY
          </label>
          <textarea
            className="brainstorm-summary-textarea"
            placeholder="Typed summary here..."
            readOnly={true}
            value=""
          />
        </div>
        <div
          className="brainstorm-form-menu"
        >
          <button
            className="cancel-btn"
            onClick={[Function]}
            type="button"
          >
            cancel
          </button>
          <button
            className="clean-btn"
            onClick={[Function]}
            type="button"
          >
            <img
              alt="broom-icon"
              className="broom"
              src="broom.svg"
            />
          </button>
          <button
            className="start-btn"
            disabled={true}
            onClick={[Function]}
            type="button"
          >
            start
          </button>
        </div>
      </form>
    </BrainstormForm>
  </Router>
</MemoryRouter>
`;
